src/commands/build/upload.rs

```diff
@@ -106,25 +107,55 @@ pub fn make_command(command: Command) -> Command {
                 .long("release-notes")
                 .help("The release notes to use for the upload.")
         )
+        .arg(
+            Arg::new("git_metadata")
+                .long("git-metadata")
+                .num_args(0..=1)
+                .default_missing_value("true")
+                .value_parser(clap::value_parser!(bool))
+                .help("Controls whether to collect and send git metadata (branch, commit, etc.). \
+                    Use --git-metadata to force enable, --git-metadata=false to force disable. \
+                    If not specified, git metadata is automatically collected only when running in a CI environment.")
+        )
 }
 
-pub fn execute(matches: &ArgMatches) -> Result<()> {
-    let config = Config::current();
-    let path_strings = matches
-        .get_many::<String>("paths")
-        .expect("paths argument is required");
+/// Holds git metadata collected for build uploads.
+#[derive(Debug, Default)]
+struct GitMetadata {
+    head_sha: Option<Digest>,
+    vcs_provider: String,
+    head_repo_name: String,
+    head_ref: String,
+    base_ref: String,
+    base_repo_name: String,
+    base_sha: Option<Digest>,
+    pr_number: Option<u32>,
+}
 
+/// Collects git metadata from arguments and VCS introspection.
+///
+/// When `auto_collect` is false, only explicitly provided values are collected;
+/// automatic inference from git repository and CI environment is skipped.
+fn collect_git_metadata(matches: &ArgMatches, config: &Config, auto_collect: bool) -> GitMetadata {
```

Comment:
i added this `auto_collect` which was not in the previous implementation. this is so that we can still use the command-line provided values (like `--head-ref foo` when `--git-metadata=false`)

Reply:
Yeah this is a good point. @szokeasaurusrex, is there some fancy Rust feature to help with the nesting here?